### PR TITLE
`Development`: Add script to generate a code coverage table automatically

### PR DESCRIPTION
**Pull request by FelixTJDietrich**
#### General

- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).

### Motivation and Context

Filling out the code coverage table of PRs has been very time consuming and cumbersome, especially for larger pull requests. Some automation in that regard would save lots of time and ensure that the code coverage table is complete and correct.

@Mtze requested this script to be added to the Artemis repository in the `supporting_scripts` top level directory, which as it turns out does not exist.

### Description

Added a `supporting_scripts` top level directory with a `generate_code_cov_table` subdirectory containing the python script to automatically generate a code coverage table from the Bamboo build artefacts.

The script requires you to provide Bamboo login credentials and by default compares the changed files between a selected branch head (default: checked out branch) and its base commit (branching point from `origin/develop`). Afterwards, the Bamboo build plan artefacts (default: latest build number) are crawled for each changed file and the line coverage is extracted. Finally, a table according to the PR template is created and stored in the clipboard.

Additionally, added a comment in the `PULL_REQUEST_TEMPLATE.md` hinting at the existence of this script.

### Steps for Testing

1. Follow the setup steps in the `supporting_scripts/generate_code_cov_table/README.md`
2. Example command: `python generate_code_cov_table.py --branch-name feature/programming-exercise/exercise-options-consolidation`
3. Try out different branches (both `client_tests` and `server_tests` have to pass for that branch, use `--info` for more information on how to change the Bamboo build number etc.)
4. Enable `--verbose` flag and check if skipped files and included files are correct
5. Past the table into a markdown viewer and check if it is correct

#### Code Review
- [x] Code Review 1
- [x] Code Review 2

#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

**Output of `python generate_code_cov_table.py --branch-name origin/chore/remove-server-api-url`:**
image

**Example of a generated table:**
image

### CODE REVIEW

**Review by pal03377 — CHANGES_REQUESTED** [review]
Very cool tool + code looks mostly very good, well done! 🎉 I also tested it with on an unrelated branch to verify the functionality.

However, I think for this script to be discoverable by new developers, it should be mentioned somewhere in the Artemis documentation. This should be added to this PR. Only then it really makes sense to have it here and not be forgotten sometime in the future.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -84,6 +84,7 @@ Prerequisites:
 <!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
 <!-- The line coverage must be above 90% for changes files and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
 <!-- Note: Use the table below and confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
+<!--       You can use `supporting_script/generate_code_cov_table/generate_code_cov_table.py` to automatically generate one from the corresponding Bamboo build plan artefacts. -->
```
[the remark this answers is not in the document]
Comment (FelixTJDietrich) [reply .github/PULL_REQUEST_TEMPLATE.md:87]: Maybe we can detect that automatically in the future and remove them. But I will leave it at that now

File: supporting_scripts/generate_code_cov_table/generate_code_cov_table.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+def environ_or_required(key):
+    # https://stackoverflow.com/a/45392259/4306257
+    return (
+        {'default': os.environ.get(key)} if os.environ.get(key)
+        else {'required': True}
+    )
+
+def get_latest_build_id(username, password, branch_name):
+    url = f"{base_url}/rest/api/latest/plan/{project_key}-{plan_key}/branch?max-results=100"
```
[the remark this answers is not in the document]
Comment (FelixTJDietrich) [reply supporting_scripts/generate_code_cov_table/generate_code_cov_table.py]: Fixed, I'm using a different endpoint now, this is no longer necessary

File: supporting_scripts/generate_code_cov_table/generate_code_cov_table.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+def filter_files(file_names, verbose):
+    client_file_names = []
+    server_file_names = []
+
+    for file_name in file_names:
+        if file_name.startswith("src/main/webapp/app"):
+            if file_name.endswith(".ts") and "module.ts" not in file_name:
+                client_file_names.append(file_name[len("src/main/webapp/"):])
+            elif verbose:
+                print(f"Skipping {file_name}")
```
[the remark this answers is not in the document]
Comment (FelixTJDietrich) [reply supporting_scripts/generate_code_cov_table/generate_code_cov_table.py]: I changed it and use the logging library now :)

**Comment by FelixTJDietrich** [comment]
Thanks @pal03377 for the detailed review, as always :) I implemented all the improvements. I think for discoverability, adding it to the PR template is enough.

**Review by pal03377 — DISMISSED** [review]

**Review by Hialus — CHANGES_REQUESTED** [review]
I tried this script out and it worked very well. Though it could use some error handling.
For example, if the tests are not done yet or if the tests failed your script currently simply throws an `IndexError: list index out of range` without any further information why it failed.

I also really don't like the fact that I have to provide my TUM credentials. At the very least you should support PATs for this. 
Though the best way would be to not use Bamboo at all. GitHub also executes the tests and records the coverage. Exporting this coverage as an artifact of the job would allow you to use this instead. This may not even require any PATs, though support for them would be nice.
This way we could then also integrate this script into the testing workflow and maybe add the option to add and update this table automatically. Though this would obviously be something for another PR.

A last point: The code currently does not adhere to [PEP-8](https://peps.python.org/pep-0008/). I think it would be important to follow this well-established style guide. E.g. There should be two empty lines before a function definition.

**Comment by FelixTJDietrich** [comment]
> I tried this script out and it worked very well. Though it could use some error handling. For example, if the tests are not done yet or if the tests failed your script currently simply throws an `IndexError: list index out of range` without any further information why it failed.
> 
> I also really don't like the fact that I have to provide my TUM credentials. At the very least you should support PATs for this. Though the best way would be to not use Bamboo at all. GitHub also executes the tests and records the coverage. Exporting this coverage as an artifact of the job would allow you to use this instead. This may not even require any PATs, though support for them would be nice. This way we could then also integrate this script into the testing workflow and maybe add the option to add and update this table automatically. Though this would obviously be something for another PR.
> 
> A last point: The code currently does not adhere to [PEP-8](https://peps.python.org/pep-0008/). I think it would be important to follow this well-established style guide. E.g. There should be two empty lines before a function definition.

Thanks, I will implement the feedback. I initially tried the PAT approach, but somehow that only works for the Bamboo REST requests and not for the artifacts. For example, I can't manage to request the following url using the PAT, only user auth works for me: `https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4773-12/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/ExamUserRepository.html`. With a PAT I get redirected to log in.

I also just noticed that the script only works for branches that have been checked out before. I will try to fix that.

**Comment by Hialus** [comment]
> Thanks, I will implement the feedback. I initially tried the PAT approach, but somehow that only works for the Bamboo REST requests and not for the artifacts. For example, I can't manage to request the following url using the PAT, only user auth works for me: `https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4773-12/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/students/exam-students.component.ts.html`. With a PAT I get redirected to log in.

You'd probably have to use the Bamboo REST API to do this, which should also give you everything you need.

> Regarding getting the code coverage through GitHub, I tried to automate the steps that I was doing manually using Bamboo. How would you get the coverage with GitHub? I saw the `codecov.io` reports [example](https://app.codecov.io/github/ls1intum/Artemis/commit/fae94f5a2ac30888b2bf00a3aa7f0e45642d4a5a/indirect-changes), but I don't know what to do there.

If you want we can work on the GitHub side together. I could adjust the workflows and then you could simply use the GitHub REST API, which is fairly easy to use. Just write me on Slack if you want to do this 🙂

**Comment by FelixTJDietrich** [comment]
I improved the error handing, PEP8 style, and better user feedback. Also `origin/*` as branch name is now supported and there is a progress bar now and clearer instructions :)

As for user auth, we sadly have to keep it, but I added a security risk note in the documentation, and you always have the option to get safely prompted for it. The reason for PAT not working is that I somehow cannot download the artifacts and have to request the generated html report without the REST API, I tried for a bit and could not figure it out. I'm using the REST API for the rest of the script though, but I thought it does not make sense to have both PAT and user auth. 

As for having it work with GitHub workflows, I would leave this to a future PR.

**Review by DominikRemo — APPROVED** [review]
Works well for me 🚀

**Review by pal03377 — APPROVED** [review]
Works well, progress bar is really cool & code looks good 💯

**Review by rriyaldhi — COMMENTED** [review]
Tested using this PR: [#6495](https://github.com/ls1intum/Artemis/pull/6495), there is 1 row that shows N/A:
image

The other rows are correct.

**Comment by FelixTJDietrich** [comment]
> Tested using this PR: [#6495](https://github.com/ls1intum/Artemis/pull/6495), there is 1 row that shows N/A: image
> 
> The other rows are correct.

Fixed thanks! The issue was a regex one. I'm now doing it more reliably.

**Review by rriyaldhi — DISMISSED** [review]
Tested using this PR: [#6495](https://github.com/ls1intum/Artemis/pull/6495) and works as expected.

**Review by laadvo — DISMISSED** [review]
Retested after the testing session, the N/A%s from before are now gone.
Before:
image
Now:
image

**Review by pellumbbaboci — APPROVED** [review]
Tested with different branches on Mac Pro , works perfectly 👍

File: supporting_scripts/generate_code_cov_table/generate_code_cov_table.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
+import json
+import re
```
Comment (pal03377) [inline supporting_scripts/generate_code_cov_table/generate_code_cov_table.py:2]: Unused import

```suggestion
```

File: supporting_scripts/generate_code_cov_table/generate_code_cov_table.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        else {'required': True}
+    )
+
+
+def get_latest_build_id(username, password, branch_name):
+    bamboo_branch_name = branch_name.replace("origin/", "").replace("/", "-")
+    url = f"{base_url}/rest/api/latest/plan/{project_key}-{plan_key}/branch/{bamboo_branch_name}.json"
+    response = requests.get(url, auth=(username, password))
+    if response.status_code != 200:
+        logging.info(f"Branch {branch_name} not found in Bamboo")
```
Comment (pal03377) [inline supporting_scripts/generate_code_cov_table/generate_code_cov_table.py:40]: This error might be a bit misleading, e.g. if Bamboo sends a response of 500 or 503. Then you would get a "branch not found" error although the issue is another. Also, I think this should be a `logging.error` here.

File: supporting_scripts/generate_code_cov_table/generate_code_cov_table.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+
+def get_code_cov_report_url(username, password, key):
+    url = f"{base_url}/rest/api/latest/result/{key}.json?expand=artifacts"
+    response = requests.get(url, auth=(username, password))
+    data = json.loads(response.content)
+    try:
+        return data["artifacts"]["artifact"][1]["link"]["href"] # second artifact is the code coverage report
+    except IndexError:
+        logging.info(f"Code coverage report not found for {key}, please wait for the tests to finish and try again")
```
Comment (pal03377) [inline supporting_scripts/generate_code_cov_table/generate_code_cov_table.py:61]:
```suggestion
        logging.warning(f"Code coverage report not found for {key}, please wait for the tests to finish and try again")
```

**Review by pal03377 — DISMISSED** [review]
Tested the script again. For deleted files, the coverage is `N/A`, which I think is ok.

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [app.constants.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4847-71/artifact/shared/Coverage-Report-Client-Tests/app/app.constants.ts.html) | 100% | ✅❌ |
| [assessment-locks.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4847-71/artifact/shared/Coverage-Report-Client-Tests/app/assessment/assessment-locks/assessment-locks.component.ts.html) | 90.32% | ✅❌ |
| [translation.config.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4847-71/artifact/shared/Coverage-Report-Client-Tests/app/core/config/translation.config.ts.html) | 100% | ✅❌ |
| [declarations.d.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4847-71/artifact/shared/Coverage-Report-Client-Tests/app/declarations.d.ts.html) | N/A% | ❌ |
| [environment.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4847-71/artifact/shared/Coverage-Report-Client-Tests/app/environments/environment.ts.html) | 100% | ✅❌ |
| [test-case-distribution-chart.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4847-71/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts.html) | 97.46% | ✅❌ |
| [navbar.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4847-71/artifact/shared/Coverage-Report-Client-Tests/app/shared/layouts/navbar/navbar.component.ts.html) | 90.87% | ✅❌ |
| [statistics-average-score-graph.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4847-71/artifact/shared/Coverage-Report-Client-Tests/app/shared/statistics-graph/statistics-average-score-graph.component.ts.html) | 94.19% | ✅❌ |



The code looks good to me except for minor comments.

**Review by rriyaldhi — APPROVED** [review]

**Review by DominikRemo — DISMISSED** [review]
Works well for me, reapproved 🚀

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -84,6 +84,8 @@ Prerequisites:
...
+<!--       Remove rows with only trivial changed from the table. -->
```
Comment (pal03377) [inline .github/PULL_REQUEST_TEMPLATE.md:88]: *changes

File: supporting_scripts/generate_code_cov_table/.gitignore [hunk 8]
```diff
@@ -0,0 +1,2 @@
+.env
+venv
```
Comment (pal03377) [inline supporting_scripts/generate_code_cov_table/.gitignore:2]: Maybe also add `__pycache__` and `*.pyc` files here? Maybe also `.DS_Store`

**Review by pal03377 — APPROVED** [review]
code looks good, just 2 more minor suggestions here

**Review by pal03377 — APPROVED** [review]

**Review by pal03377 — DISMISSED** [review]

**Review by DominikRemo — DISMISSED** [review]

**Review by Hialus — CHANGES_REQUESTED** [review]
Tried it again and providing the password over the CLI doesn't work as it should:
- The help dialog says `--password PASSWORD   Bamboo password (optional, will be prompted if not provided)`
- I run `python3 generate_code_cov_table.py --username ge49juk --branch-name "bugfix/improve-consistency-check"`
- I get the error `generate_code_cov_table.py: error: the following arguments are required: --password`
- What I would expect: If `--password` is not provided I can input it in the CLI

Note: I did not set up a .env file.

**Comment by FelixTJDietrich** [comment]
> Tried it again and providing the password over the CLI doesn't work as it should:
> 
> * The help dialog says `--password PASSWORD   Bamboo password (optional, will be prompted if not provided)`
> * I run `python3 generate_code_cov_table.py --username ge49juk --branch-name "bugfix/improve-consistency-check"`
> * I get the error `generate_code_cov_table.py: error: the following arguments are required: --password`
> * What I would expect: If `--password` is not provided I can input it in the CLI
> 
> Note: I did not set up a .env file.

Thanks for finding that. I tested it without .env file but forgot that I still had my password in the environment. Now it is working again :| 

Also there is a windows encoding issue with the emojis when using the default copy to clipboard option. Hopefully that is fixed now 🤞 but I cannot test it.

**Review by DominikRemo — APPROVED** [review]
Reapproved

**Review by bassner — APPROVED** [review]

**Pull request merged**